0867/Original-Submission/Src/quadlog-gampsi-slim/man/dgamma.man
.TH DGAMMA 3 "04 August 2000" "Version 1.00"
.\" WARNING: This file was produced automatically from file common/dgamma.f
.\" by fortran-to-man-page.awk on Sun Dec 31 09:02:05 MST 2000.
.\" Any manual changes will be lost if this file is regenerated!
.SH NAME
dgamma \- Double-precision Gamma(x)
.\"=====================================================================
.SH SYNOPSIS
Fortran (77, 90, 95, HPF):
.RS
.B f77
.I "[ flags ] file(s) .\|.\|. -L/usr/local/lib -lgjl"
.RS
.nf
.B "DOUBLE PRECISION FUNCTION dgamma(x)"
.B "DOUBLE PRECISION    x"
.fi
.RE
.RE
C (K&R, 89, 99), C++ (98):
.RS
.B cc
.I "[ flags ] -I/usr/local/include file(s) .\|.\|. -L/usr/local/lib -lgjl"
.br
Use
.RS
.B "#include <gampsi.h>"
.RE
to get this prototype:
.RS
.B "fortran_double_precision dgamma(const fortran_double_precision * x_);"
.RE
.RE
.PP
NB: The definition of C/C++ data types
.B fortran_
.IR xxx ,
and the mapping of Fortran external names to C/C++ external names,
is handled by the C/C++ header file.  That way, the same function
or subroutine name can be used in C, C++, and Fortran code,
independent of compiler conventions for mangling of external
names in these programming languages.
.PP
Last code modification: 04-Aug-2000
.\"=====================================================================
.SH DESCRIPTION
Compute and return the value of the Gamma(\fBx\fP\&) function for
double-precision \fBx\fP\&.
.PP
This code correctly handles the case where \fBx\fP\& is NaN, for which
Gamma(NaN) is a NaN. and the case where \fBx\fP\& is sufficiently large
and positive, or takes one of the integer values 0, -1, -2, -3,
\&.\|.\|., for which Gamma(\fBx\fP\&) = +Infinity, a special value in IEEE 754
arithmetic.
.PP
This code is substantially similar to that given by
.PP
.RS
.nf
W. J. Cody
Algorithm 715: SPECFUN --- A Portable FORTRAN
Package of Special Function Routines and Test Drivers
ACM Trans. Math. Software 19(1) 22--32, March 1993.
.fi
.RE
.PP
but has been augmented for support of IEEE 754 arithmetic.
.\"=====================================================================
.SH "SEE ALSO"
.BR dpsi (3),
.BR dpsiln (3),
.BR gamma (3),
.BR psi (3),
.BR psiln (3),
.BR qgamma (3),
.BR qpsi (3),
.BR qpsiln (3).
.\"=====================================================================
.SH AUTHORS
The algorithms and code are described in detail in
the paper
.RS
.I "Algorithm xxx: Quadruple-Precision Gamma(x) and psi(x) Functions for Real Arguments"
.RE
in ACM Transactions on Mathematical Software,
Volume ??, Number ??, Pages ????--???? and
????--????, 2001, by
.RS
.nf
Nelson H. F. Beebe
Center for Scientific Computing
University of Utah
Department of Mathematics, 110 LCB
155 S 1400 E RM 233
Salt Lake City, UT 84112-0090
Tel: +1 801 581 5254
FAX: +1 801 581 4148
Email: \fCbeebe@math.utah.edu\fP, \fCbeebe@acm.org\fP, \fCbeebe@computer.org\fP
WWW URL: \fChttp://www.math.utah.edu/~beebe\fP
.fi
.RE
and
.RS
.nf
James S. Ball
University of Utah
Department of Physics
Salt Lake City, UT 84112-0830
USA
Tel: +1 801 581 8397
FAX: +1 801 581 6256
Email: \fCball@physics.utah.edu\fP
WWW URL: \fChttp://www.physics.utah.edu/people/faculty/ball.html\fP
.fi
.RE
.\"==============================[The End]==============================
.\"=====================================================================
.\" This is for GNU Emacs file-specific customization:
.\" Local Variables:
.\" fill-column: 50
.\" End:


Changes
April 22 2021
   * Added 1013-1014 from Volume 47, number 1, (January 2021)
   * Added 1012 from Volume 46, number 4, (November 2020)
October 19 2020
   * Added 1011 from Volume 46, number 3, (September 2020)
August 7 2020
   * Added 1003-1006 from Volume 46, number 1, (April 2020)
   * Added 1007-1010 from Volume 46, number 2, (June 2020)
January 12 2020
   * Updated 999 at the request of the author -- updated user manual supplied.
January 7 2020
   * Added 999-1002 from Volume 45, number 4, (December 2019)
October 13 2019
   * Added 995-998 from Volume 45, number 3, (August 2019)
August 8 2019
   * Added 993 and 994 from Volume 45, number 2, (June 2019)
August 7 2019
   * Added 991 and 992 from Volume 45, number 1, (February 2019)
February 1 2019
   * Added 989 and 990 from Volume 44, number 3, (August 2018)
May 8 2018
   * Added 987 and 988 from Volume 44, number 3, (April 2018)
October 13 2017
   * Added 986 from Volume 44, number 2, (October 2017)
   * Corrected publication month for Algs 979-985 (now Oct).
September 30 2017
   * Added 979-985 from Volume 44, number 2, (September 2017)
   * Added 975-978 from Volume 44, number 1, (July 2017)
April 21 2017
   * Added 973, 974 and Remark on 936 from Volume 43, number 4 
     (March 2017)
February 3 2017
   * Added 968-972 from Volume 43, number 3 (January 2017)
September 16 2016
   * Added 965 from Volume 43, number 1 (August 2016)
   * Added 966, 967 from Volume 43, number 2 (September 2016)
   * Removed files 620 and 620.bib (all other algorithm files
     are only available as compressed files.
   * Replaced index.gz and index.html.gz by .zip files
September 14 2016:
   * Added 100.zip and 101.zip. These files are identical as
   the two algorithms are inter-related and the driver program
   uses both codes.
August 30 2016:
   * Added 963-964 from Volume 42, number 4 (July 2016)
July 23 2016
   * Updated contact information on CALGO web pages.
June 19 2016:
   * Added 960-962 from Volume 42, number 3 (June 2016)
   * Updated 916 to include the code submitted with Remark 
     in Volume 42, number 3 (Jun 2016)
June 15 2016:
   * Added 957 from Volume 42, number 1 (February 2016)
   * Added 958-959 from Volume 42, number 2 (June 2016)
May 30 2016
   * Added 955-956 from Volume 42, number 1 (February 2016)
May 27 2016:
   * Added new code supplied as Remark to Algorithm 673
     appearing in Volume 42, number 1.
December 22 2015:
   * Added 947-954 from Volume 41, numbers (1-4).
August 29 2015:
   * Mended a number of odd/broken compressed files that were on
     the CALGO site:
       755.gz, 756.gz, 823.gz replaced by zip files
       709.gz, 709a.gz, 709b.gz replaced by 709.zip
       573.gz, 573d.gz, 573s.gz replaced by 573.zip
       358.gz, 805.gz, 830.gz, 832.zip all updated
       620.zip added -- this contains all the 620 components.
September 26 2014
   * Updated 897 following accepted remark.
September 1 2014
   * Added 943-946 from Volume 40, number 4 (June 2014)
May 17 2014
   * Fixed problem with Algorithm 917 missing from indexes.
May 5 2014
   * Added 935-938 from Volume 40, number 2 (February 2014)
   * Added 939-942 from Volume 40, number 3 (April 2014)
October 9 2013
   * Added 929, 930 from volume 39, number 4 (July 2013)
   * Added 931-934 from volume 40, number 1 (October 2013)
July 18 2013
   * Added 928 from volume 39, number 3 (May 2013)
   * Updated 905 from code supplied by author thackerw@winthrop.edu
March 3 2013
   * Added 926-927 from volume 39, number 2 (February 2013)
January 25 2013
   * Replaced 923 after editing Licence file -- author request
January 3 2013
   * Added 924-925 from volume 39, number 1 (November 2012)
September 10 2012
   * Added 920-923 from volume 38, number 4 (August 2012)
June 17 2012
   * Added 917-919 from volume 38, number 3 (April 2012)
March 14 2012
   * Corrected typo in index files (Alg 916)
February 29 2012
   * Replaced 916 -- old version of code uploaded in error.
February  5 2012
   * Added 916 from Volume 38, number 2 (December 2012)
December 30 2011
   * Replaced 905 due to corrupt blas.f file in the original (pointed
     out by Ronald.M.Bloom@aero.org).
December 7 2011
   * Added 913-915 from Volume 38, number 1 (November 2011)
   * Updates 778 as per Remark in Volume 38, No. 1
April 12 2011
   * Added 909-911 from volume 37, number 4 (April 2011)
October 24 2011
   * Changed Algorithm 419 in line with changes from Remark on Algorithm 419 by
     David H. Withers CACM (March 1974) Vol 17 No 3 p. 157
     Pointed out by j.a.horsmeier@wanadoo.nl (Jos)
October 11 2010
   * Added 904-908 from volume 37, number 3 (September 2010)
May  20 2010
   * Added 902, 903 from volume 37, number 2 (April 2010)
Feb  27 2010
   * Added 898, 899, 900, 901 from volume 37, number 1 (January 2010)
July 29 2009
   * Added 895, 896, 897 from volume 36, number 3 (July 2009)
Apr  6 2009
   * Added 894 from volume 36, number 2 (March 2009)
Mar  30 2009
   * Added 891, 892, 893 from volume 36, number 1 (March 2009)
Feb  12 2009
   * Added 890 from volume 35, number 4 (February 2009)
Jan   6 2009
   * Added 885-889 from volume 35, number 3 (October 2008)
Sept  4 2008
   * Added 881-884 from volume 35, number 2 (July 2008)
July 25 2008
   * Added 876-877 from volume 34, number 4 (July 2008)
   * Added 878-880 from volume 35, number 1 (July 2008)
June 11 2008
   * Added 873-875 from volume 34, number 2 and 3 (March and May 2008)
Feb   8 2008
   * Added 867-869 from volume 33, numbers 3 and 4 (August 2007)
   * Added 870-872 from volume 34, number 1 (January 2008)
Aug   3 2007
   * Added 863-866 from volume 33, numbers 1 and 2.
July  3 2007
   * Updated 620 to include GAMS classifications to algorithms 814-862 inc.
Jan  10 2007
   * Updated 644 as a result of an accepted remark by Kodama
Jan   5 2007
   * Added 857-862 from volume 32, 4 (December 2006)
Nov  21 2006
   * Added 358 -- donated by John Burkardt <burkardt@csit.fsu.edu>.
Oct   5 2006
   * Added 854 from volume 32, 2 (June 2006)

867/Src/quadlog-gampsi-slim/common/dgamma.f
      DOUBLE PRECISION FUNCTION  dgamma(x)
************************************************************************
*     (Double-precision Gamma(x))
*     Compute and return the value of the Gamma(x) function for
*     double-precision x.
*
*     This code correctly handles the case where x is NaN, for which
*     Gamma(NaN) is a NaN. and the case where x is sufficiently large
*     and positive, or takes one of the integer values 0, -1, -2, -3,
*     ..., for which Gamma(x) = +Infinity, a special value in IEEE 754
*     arithmetic.
*
*     This code is substantially similar to that given by
*
*         W. J. Cody
*         Algorithm 715: SPECFUN --- A Portable FORTRAN
*         Package of Special Function Routines and Test Drivers
*         ACM Trans. Math. Software 19(1) 22--32, March 1993.
*
*     but has been augmented for support of IEEE 754 arithmetic.
*     (04-Aug-2000)
************************************************************************
*     FOREWORD: This function is derived from code available on the
*     World-Wide Web at ftp://ftp.netlib.org/toms/715, corresponding to
*     the article
*
*         W. J. Cody
*         Algorithm 715: SPECFUN --- A Portable FORTRAN
*         Package of Special Function Routines and Test Drivers
*         ACM Trans. Math. Software 19(1) 22--32 March 1993.
*
*     Additional code has been added to handle the cases of x = NaN, for
*     which Gamma(NaN) = NaN, and large x, and x = 0, -1, -2, -3, ...,
*     for which Gamma(x) = +Infinity, a special value in IEEE 754
*     arithmetic.
*
*     The code has been prettyprinted to match the conventions of the
*     quadrature package in which it is distributed, and the
*     single-precision code embedded in comments has been discarded.
*
*     The constants eps, xbig, xminin, and xinf below have been given
*     values suitable for IEEE 754 double-precision arithmetic.  This is
*     applicable on all current Intel, Motorola, and RISC CPUs, and even
*     on IBM S/390 mainframes since 1999.  Older Compaq/DEC VAX, Convex,
*     Cray, and IBM S/3x0 systems will require changes in those values.
*
************************************************************************
*
*     This routine calculates the gamma function for a real argument x.
*
*     Computation is based on an algorithm outlined in W. J. Cody, ``An
*     overview of software development for special functions'', Lecture
*     Notes in Mathematics, 506, Numerical Analysis Dundee, 1975,
*     G. A. Watson (Ed.), Springer Verlag, Berlin, 1976.
*
*     The program uses rational functions that approximate the gamma
*     function to at least 20 significant decimal digits.  Coefficients
*     for the approximation over the interval (1,2) are unpublished.
*     Those for the approximation for x .ge. 12 are from Hart, et. al.,
*     Computer Approximations, Wiley and Sons, New York, 1968.  Lower
*     order approximations can be substituted for these on machines with
*     less precise arithmetic.
*
************************************************************************
************************************************************************
*
*     Explanation of machine-dependent constants:
*
*     eps    - the smallest positive floating-point number such that
*              1.0 + eps .gt. 1.0
*     xbig   - the largest argument for which Gamma(x) is
*              representable in the machine, i.e., the solution to the
*              equation
*                      Gamma(xbig) = xinf.
*     xminin - the smallest positive floating-point number such that
*              1/xminin is machine representable.
*     xinf   - the largest machine representable floating-point number.
*
*     Approximate values for some important machines are:
*
*             IBM/195    CDC/7600  UNIVAC/1108   VAX 11/780 (UNIX)
*              (d.p.)  (s.p.,rndg)    (d.p.)     (s.p.)     (d.p.)
*
*     eps     2.221d-16  3.553e-15   1.735d-18   5.961e-08  1.388d-17
*     xbig    57.574     177.802     171.489     34.844     34.844
*     xminin  1.382d-76  3.132e-294  1.113d-308  5.883e-39  5.883d-39
*     xinf    7.23d+75   1.26e+322   8.98d+307   1.70e+38   1.70d+38
*
************************************************************************
************************************************************************
*
*
*     Error returns:
*
*      The program returns the value xinf for singularities or
*      when overflow would occur.  The computation is believed
*      to be free of underflow and overflow.
*
*
*
*     Other subprograms required (single precision version):
*
*         alog,exp,float,ifix,sin
*
*     Other subprograms required (double precision version):
*
*         dble,dexp,dlog,dsin,float,ifix,sngl
*
*
*
*      Author: W. J. Cody
*              Applied Mathematics Division
*              Argonne National Laboratory
*              Argonne, IL 60439
*
*      Latest modification: May 18, 1982
*
************************************************************************
*
*     Intrinsic functions
*
      INTRINSIC           dcos,        dexp,        dint,        dlog
      INTRINSIC           dsin,        idint,       mod
*
*     Built-in functions
*
      DOUBLE PRECISION    dcos,        dexp,        dint,        dlog
      DOUBLE PRECISION    dsin
*
      INTEGER             idint,       mod
*
*     External functions
*
      EXTERNAL            dfloat,      dinf,        dnan,        isdnan
*
      DOUBLE PRECISION    dfloat,      dinf,        dnan
*
      LOGICAL             isdnan
*
*     Parameter variables
*
      DOUBLE PRECISION    four
      PARAMETER           (four = 4.0d+00)
*
      DOUBLE PRECISION    half
      PARAMETER           (half = 0.5d0)
*
      DOUBLE PRECISION    one
      PARAMETER           (one = 1.0d0)
*
*     DOUBLE PRECISION    pi
*     PARAMETER           (pi =
*    X    3.14159265358979323846264338327950288419716939937511d+00)
*
*     piov4 = pi / 4
*
      DOUBLE PRECISION    piov4
      PARAMETER           (piov4 =
     X    7.853981633974483096156608458198757210492923498437764552d-01)
*
      DOUBLE PRECISION    two
      PARAMETER           (two = 2.0d0)
*
      INCLUDE 'dl2pi2.inc'
*
      INCLUDE 'deps.inc'
*
      INCLUDE 'dxinf.inc'
*
      INCLUDE 'dxbig.inc'
*
      DOUBLE PRECISION    cutoff
*     PARAMETER           (cutoff = 12.0d0)
      PARAMETER           (cutoff = xbig + one)
*
      INCLUDE 'dxmin.inc'
*
*     Maple V5.1:
*         Digits := 25;
*         printf("%.25e\n", 2^(-1022)):
*             2.2250738585072013830902330e-308
*         printf("%.25e\n", 2^1022):
*             4.4942328371557897693232630e+307
*
*     Thus, the reciprocal is representable, since it is less than xinf.
*     There are smaller denormalized values whose reciprocals are
*     representable, the smallest being
*
*         printf("%.25e\n", 1/((2 - 2^(-53)) * 2^1023)):
*             5.5626846462680037665166100e-309
*
*     However, some architectures carry severe performance penalties for
*     denormalized values, so we avoid using them.
*
      DOUBLE PRECISION    xminin
      PARAMETER           (xminin = xmin)
*
      DOUBLE PRECISION    zero
      PARAMETER           (zero = 0.0d0)
*
*     Argument variables
*
      DOUBLE PRECISION    x
*
*     Local variables
*
      DOUBLE PRECISION    c(7),        fact,        p(8),        q(8)
      DOUBLE PRECISION    result,      sinval,      sum,         u
      DOUBLE PRECISION    v,           w,           xden,        xnum
      DOUBLE PRECISION    y,           y1,          yfrac,       yint
      DOUBLE PRECISION    ysq,         z
*
      INTEGER             i,           n,           umod8
*
      LOGICAL             parity
*
      DATA p/ -1.71618513886549492533811d+0,
     X    2.47656508055759199108314d+1, -3.79804256470945635097577d+2,
     X    6.29331155312818442661052d+2, 8.66966202790413211295064d+2,
     X    -3.14512729688483675254357d+4, -3.61444134186911729807069d+4,
     X    6.64561438202405440627855d+4/
      DATA q/ -3.08402300119738975254353d+1,
     X    3.15350626979604161529144d+2, -1.01515636749021914166146d+3,
     X    -3.10777167157231109440444d+3, 2.25381184209801510330112d+4,
     X    4.75584627752788110767815d+3, -1.34659959864969306392456d+5,
     X    -1.15132259675553483497211d+5/
      DATA c/ -1.910444077728d-03, 8.4171387781295d-04,
     X    -5.952379913043012d-04, 7.93650793500350248d-04,
     X    -2.777777777777681622553d-03, 8.333333333333333331554247d-02,
     X    5.7083835261d-03/
*
      parity = .FALSE.
      fact = one
      n = 0
      y = x
      IF (isdnan(x)) THEN
*-----------------------------------------------------------------------
*          Argument is a NaN, so Gamma(NaN) is too.
*-----------------------------------------------------------------------
           dgamma = dnan()
           RETURN
      ELSE IF (y .LE. zero) THEN
*-----------------------------------------------------------------------
*         Argument is zero or negative.  Gamma(x) is finite, except at
*         the poles where x = 0, -1, -2, -3, ...
*-----------------------------------------------------------------------
          y = -x
          yint = dint(y)
          yfrac = y - yint
          IF (yfrac .NE. zero) THEN
              parity = (yint .NE. dint(yint*half)*two)
*-----------------------------------------------------------------------
*             We want to set fact = -pi/dsin(pi*yfrac).  However, pi
*             has two leading zero bits in IBM S/360 arithmetic, so
*             compute this more accurately as fact = -4*(pi/4) /
*             sin((pi/4)*(4*yfrac)), and reduce this further by
*             4*yfrac = w = u (integer) + v (fraction), where w, u,
*             and v are all nonnegative and small, producing eight
*             cases involving arguments with v and 1-v (which does not
*             suffer from subtraction loss).
*-----------------------------------------------------------------------
              w = four * yfrac
              u = dint(w)
              v = w - u
              umod8 = mod(idint(u),8)
              IF (umod8 .eq. 0) THEN
                   sinval = dsin(piov4 * v)
              ELSE IF (umod8 .eq. 1) THEN
                   sinval = dcos(piov4 * (one -v))
              ELSE IF (umod8 .eq. 2) THEN
                   sinval = dcos(piov4 * v)
              ELSE IF (umod8 .eq. 3) THEN
                   sinval = dsin(piov4 * (one - v))
              ELSE IF (umod8 .eq. 4) THEN
                   sinval = -dsin(piov4 * v)
              ELSE IF (umod8 .eq. 5) THEN
                   sinval = -dcos(piov4 * (one - v))
              ELSE IF (umod8 .eq. 6) THEN
                   sinval = -dcos(piov4 * v)
              ELSE
                   sinval = -dsin(piov4 * (one - v))
              END IF
              fact = -four * (piov4 / sinval)
*OLD:         fact = (-pi)/dsin(pi*yfrac)
              y = y + one
          ELSE
*-----------------------------------------------------------------------
*             Argument is negative integer, so Gamma(x) = +/-Infinity
*-----------------------------------------------------------------------
              result = dinf()
              GO TO 400
          END IF
      END IF
*-----------------------------------------------------------------------
*     Argument is now positive, with fact and parity recording history
*     of negative argument.
*-----------------------------------------------------------------------
      IF (y .LT. eps) THEN
*-----------------------------------------------------------------------
*         Argument .LT. eps
*-----------------------------------------------------------------------
          IF (y .GE. xminin) THEN
              result = one/y
          ELSE
              result = dinf()
              GO TO 400
          END IF
      ELSE IF (y .LT. cutoff) THEN
          y1 = y
          IF (y .LT. one) THEN
*-----------------------------------------------------------------------
*             0.0 .LT. argument .LT. 1.0
*-----------------------------------------------------------------------
              z = y
              y = y + one
          ELSE
*-----------------------------------------------------------------------
*            1.0 .LT. argument .LT. cutoff, reduce argument if necessary
*-----------------------------------------------------------------------
              n = idint(y) - 1
              y = y - dfloat(n)
              z = y - one
          END IF
*-----------------------------------------------------------------------
*         Evaluate approximation for 1.0 .LT. argument .LT. 2.0
*-----------------------------------------------------------------------
          xnum = zero
          xden = one
          DO 100 i = 1, 8
              xnum = (xnum + p(i))*z
              xden = xden*z + q(i)
  100     CONTINUE
          result = xnum/xden + one
          IF (y1 .LT. y) THEN
*-----------------------------------------------------------------------
*             Adjust result for case  0.0 .LT. argument .LT. 1.0
*-----------------------------------------------------------------------
              result = result/y1
          ELSE IF (y1 .GT. y) THEN
*-----------------------------------------------------------------------
*             Adjust result for case  2.0 .LT. argument .LT. cutoff
*-----------------------------------------------------------------------
              DO 200 i = 1, n
                  result = result*y
                  y = y + one
  200         CONTINUE
          END IF
      ELSE
*-----------------------------------------------------------------------
*         Argument .GE. cutoff
*-----------------------------------------------------------------------
          IF (y .LE. xbig) THEN
*-----------------------------------------------------------------------
*             Use asymptotic series for ln(Gamma(x)).  The four
*             parts are summed in order from smallest to largest,
*             positive parts first.
*-----------------------------------------------------------------------
              ysq = y*y
              sum = c(7)
              DO 300 i = 1, 6
                  sum = sum/ysq + c(i)
  300         CONTINUE
              sum = ((y - half)*dlog(y) +(ln2pi2 + sum/y)) - y
              result = dexp(sum)
          ELSE
              result = dinf()
          END IF
      END IF
*-----------------------------------------------------------------------
*     Final adjustments and return
*-----------------------------------------------------------------------
  400 IF (parity) result = -result
      IF (fact .NE. one) result = fact/result
      dgamma = result
      END
